papa/src/main/java/papa/internal/RealInputTracker.kt

```diff
@@ -23,49 +24,84 @@ import papa.InputTracker
 import papa.SafeTrace
 import papa.internal.FrozenFrameOnTouchDetector.findPressedView
 import papa.safeTrace
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration.Companion.nanoseconds
 
 internal object RealInputTracker : InputTracker {
 
   override val motionEventTriggeringClick: DeliveredInput<MotionEvent>?
-    get() = motionEventTriggeringClickLocal.get()
+    get() = motionEventTriggeringClickLocal.get()?.input
 
   override val currentKeyEvent: DeliveredInput<KeyEvent>?
     get() = currentKeyEventLocal.get()
 
-  private val motionEventTriggeringClickLocal = ThreadLocal<DeliveredInput<MotionEvent>>()
+  private val motionEventTriggeringClickLocal = ThreadLocal<MotionEventHolder>()
   private val currentKeyEventLocal = ThreadLocal<DeliveredInput<KeyEvent>>()
 
   private val handler = Handler(Looper.getMainLooper())
 
+  class MotionEventHolder(var input: DeliveredInput<MotionEvent>) : Choreographer.FrameCallback {
+
+    private val choreographer = Choreographer.getInstance()
+
+    override fun doFrame(frameTimeNanos: Long) {
+      // We increase the counter right after the frame callback. This means we don't count a frame
+      // if this event is consumed as part of the frame we did the increment in.
+      // There's a slight edge case: if the event consumption triggered in between doFrame and
+      // the post at front of queue, the count would be short by 1. We can live with this, it's
+      // unlikely to happen unless and even is triggered from a postAtFront.
+      mainHandler.postAtFrontOfQueueAsync {
+        input = input.increaseFrameCount()
+      }
+      choreographer.postFrameCallback(this)
+    }
+
+    fun startCounting() {
+      choreographer.postFrameCallback(this)
+    }
+
+    fun stopCounting() {
+      choreographer.removeFrameCallback(this)
+    }
+  }
+
   private val listener = OnRootViewAddedListener { view ->
     view.phoneWindow?.let { window ->
       if (view.windowAttachCount == 0) {
         window.touchEventInterceptors += TouchEventInterceptor { motionEvent, dispatch ->
-          val isActionUp = motionEvent.action == MotionEvent.ACTION_UP
           // Note: what if we get 2 taps in a single dispatch loop? Then we're simply posting the
           // following: (recordTouch, onClick, clearTouch, recordTouch, onClick, clearTouch).
-          val deliveryUptimeMillis = SystemClock.uptimeMillis()
+          val deliveryUptimeNanos = System.nanoTime()
+          val isActionUp = motionEvent.action == MotionEvent.ACTION_UP
 
-          val (actionUpEvent, cookie) = if (isActionUp) {
-            val input = DeliveredInput(MotionEvent.obtain(motionEvent), deliveryUptimeMillis)
-            val cookie = deliveryUptimeMillis.rem(Int.MAX_VALUE).toInt()
-            input to cookie
+          val actionUpEventHolder = if (isActionUp) {
```

Comment:
This isn't obvious, but the change from "event" to "eventHolder" is so that we can actually replace the event within the holder. Why replace it? Because we want to increase the frame count over time, but we want to do that by swapping an immutable event, so that if we capture such event at time N and then the count gets updated at N + 1, the count update isn't reflected in the code that captured the event a time N.

Reply:
(and of course this should actually be a comment in the code)